File: go/tui/internal/resources/resources.go
package resources

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/gasdotdev/gas/tui/internal/graph"
)

type Resources struct {
	NameToConfig nameToConfig
}

func New() (*Resources, error) {
	r := &Resources{}

	configJson, err := getConfigJson()
	if err != nil {
		return nil, err
	}

	containerDir := getContainerDirPath(configJson)

	containerSubdirPaths, err := getContainerSubdirPaths(containerDir)
	if err != nil {
		return nil, err
	}

	nameToPackageJson, err := getNameToPackageJson(containerSubdirPaths)
	if err != nil {
		return nil, err
	}

	packageJsonNameToName := setPackageJsonNameToName(nameToPackageJson)

	nameToDeps := r.setNameToDeps(nameToPackageJson, packageJsonNameToName)

	nameToIndexFilePath, err := r.setNameToIndexFilePath(containerSubdirPaths)
	if err != nil {
		return nil, err
	}

	nameToIndexFileContent, err := r.setNameToIndexFileContent(nameToIndexFilePath)
	if err != nil {
		return nil, err
	}

	g := graph.New(graph.NodeToDeps(nameToDeps))

	groupToDepthToNames := g.GroupToDepthToNodes

	// namesWithInDegreesOfZero := g.NodesWithInDegreesOfZero

	// nameToIntermediates := g.NodeToIntermediates

	// depthToName := g.DepthToNode

	// nameToDepth := g.NodeToDepth

	nameToConfigData, err := r.setNameToConfigData(nameToIndexFileContent)
	if err != nil {
		return nil, err
	}

	nodeJsConfigScript, err := r.setNodeJsConfigScript(nameToConfigData, groupToDepthToNames)
	if err != nil {
		return nil, err
	}

	runNodeJsConfigScriptResult, err := r.runNodeJsConfigScript(nodeJsConfigScript)
	if err != nil {
		return nil, err
	}

	nameToConfig := r.setNameToConfig(runNodeJsConfigScriptResult)

	r.NameToConfig = nameToConfig

	return r, nil
}

type configJson map[string]interface{}

func getConfigJson() (configJson, error) {
	configPath := "gas.config.json"

	data, err := os.ReadFile(configPath)
	if err != nil {
		if os.IsNotExist(err) {
			return configJson{}, nil
		}
		return nil, fmt.Errorf("failed to read gas.config.json: %w", err)
	}

	var result configJson
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("failed to parse gas.config.json: %w", err)
	}

	return result, nil
}

type containerDirPath string

func getContainerDirPath(configJson configJson) containerDirPath {
	if dirPath, ok := configJson["resourceContainerDirPath"].(string); ok {
		return containerDirPath(dirPath)
	}

	return containerDirPath("./gas")
}

type containerSubdirPaths []string

func getContainerSubdirPaths(containerDirPath containerDirPath) (containerSubdirPaths, error) {
	entries, err := os.ReadDir(string(containerDirPath))

	if err != nil {
		return nil, fmt.Errorf("unable to read resource container dir %s", containerDirPath)
	}

	var containerSubdirPaths containerSubdirPaths

	for _, entry := range entries {
		if entry.IsDir() {
			containerSubdirPaths = append(containerSubdirPaths, filepath.Join(string(containerDirPath), entry.Name()))
		}
	}

	return containerSubdirPaths, nil
}

type nameToPackageJson map[string]*packageJson

type packageJson struct {
	Name            string            `json:"name"`
	Main            string            `json:"main"`
	Types           string            `json:"types"`
	Scripts         map[string]string `json:"scripts"`
	Dependencies    map[string]string `json:"dependencies,omitempty"`
	DevDependencies map[string]string `json:"devDependencies,omitempty"`
}

func getNameToPackageJson(containerSubdirPaths containerSubdirPaths) (nameToPackageJson, error) {
	nameToPackageJson := make(nameToPackageJson)

	for _, subdirPath := range containerSubdirPaths {
		resourceName := convertContainerSubdirPathToName(subdirPath)

		packageJsonPath := filepath.Join(subdirPath, "package.json")

		data, err := os.ReadFile(packageJsonPath)
		if err != nil {
			return nil, fmt.Errorf("unable to read file %s\n%v", packageJsonPath, err)
		}

		var packageJson packageJson
		err = json.Unmarshal(data, &packageJson)
		if err != nil {
			return nil, fmt.Errorf("unable to parse %s\n%v", packageJsonPath, err)
		}

		nameToPackageJson[resourceName] = &packageJson
	}

	return nameToPackageJson, nil
}

func convertContainerSubdirPathToName(subdirPath string) string {
	subdirName := filepath.Base(subdirPath)
	snakeCaseResourceName := strings.ReplaceAll(subdirName, "-", "_")
	screamingSnakeCaseResourceName := strings.ToUpper(snakeCaseResourceName)
	return screamingSnakeCaseResourceName
}

type packageJsonNameToName map[string]string

func setPackageJsonNameToName(nameToPackageJson nameToPackageJson) packageJsonNameToName {
	result := make(packageJsonNameToName)
	for resourceName, packageJson := range nameToPackageJson {
		result[packageJson.Name] = resourceName
	}
	return result
}

type nameToDeps map[string][]string

func (r *Resources) setNameToDeps(nameToPackageJson nameToPackageJson, packageJsonNameToName packageJsonNameToName) nameToDeps {
	result := make(nameToDeps)
	for resourceName, packageJson := range nameToPackageJson {
		var deps []string
		// Loop over source resource's package.json deps
		for dep := range packageJson.Dependencies {
			internalDep, ok := packageJsonNameToName[dep]
			// If package.json dep exists in map then it's an internal dep
			if ok {
				deps = append(deps, internalDep)
			}
		}
		result[resourceName] = deps
	}
	return result
}

type nameToIndexFilePath map[string]string

func (r *Resources) setNameToIndexFilePath(containerSubdirPaths containerSubdirPaths) (nameToIndexFilePath, error) {
	result := make(nameToIndexFilePath)

	indexFilePathPattern := regexp.MustCompile(`^_[^.]+\.[^.]+\.[^.]+\.index\.ts$`)

	for _, subdirPath := range containerSubdirPaths {
		subdirName := filepath.Base(subdirPath)
		snakeCaseResourceName := strings.ReplaceAll(subdirName, "-", "_")
		screamingSnakeCaseResourceName := strings.ToUpper(snakeCaseResourceName)

		srcPath := filepath.Join(subdirPath, "src")

		files, err := os.ReadDir(srcPath)
		if err != nil {
			return nil, err
		}

		for _, file := range files {
			if !file.IsDir() && indexFilePathPattern.MatchString(file.Name()) {
				result[screamingSnakeCaseResourceName] = filepath.Join(srcPath, file.Name())
				break
			}
		}
	}

	return result, nil
}

type nameToIndexFileContent map[string]string

func (r *Resources) setNameToIndexFileContent(nameToIndexFilePath nameToIndexFilePath) (nameToIndexFileContent, error) {
	result := make(nameToIndexFileContent)
	for name, indexFilePath := range nameToIndexFilePath {
		data, err := os.ReadFile(indexFilePath)
		if err != nil {
			return nil, fmt.Errorf("unable to read file %s\n%v", indexFilePath, err)
		}
		result[name] = string(data)
	}
	return result, nil
}

type nameToConfigData map[string]*configData

type configData struct {
	variableName string
	functionName string
	exportString string
}

func (r *Resources) setNameToConfigData(nameToIndexFileContent nameToIndexFileContent) (nameToConfigData, error) {
	result := make(nameToConfigData)

	for name, indexFileContent := range nameToIndexFileContent {
		// Config setters are imported like this:
		// import { cloudflareKv } from "@gasoline-dev/resources"
		// They can be distinguished using a camelCase pattern.
		configSetterFunctionNameRegex := regexp.MustCompile(`import\s+\{[^}]*\b([a-z]+[A-Z][a-zA-Z]*)\b[^}]*\}\s+from\s+['"]@gasoline-dev/resources['"]`)
		// This can be limited to one match because there should only
		// be one config setter per resource index file.
		configSetterFunctionName := configSetterFunctionNameRegex.FindStringSubmatch(indexFileContent)[1]

		// Configs are exported like this:
		// export const coreBaseKv = cloudflareKv({
		//   name: "CORE_BASE_KV",
		// } as const)
		exportedConfigRegex := regexp.MustCompile(`(?m)export\s+const\s+\w+\s*=\s*\w+\([\s\S]*?\)\s*(?:as\s*const\s*)?;?`)

		// It can't be assumed that text that matches the exported config
		// pattern is an exported config. A user can export non-configs
		// using the same pattern above. So we need to collect possible
		// exported configs and evaluate them later.
		possibleExportedConfigs := exportedConfigRegex.FindAllString(indexFileContent, -1)

		// This regex matches the variable name of an exported
		// config. For example, it'd match "coreBaseKv" in:
		// export const coreBaseKv = cloudflareKv({
		//   name: "CORE_BASE_KV",
		// } as const)
		possibleExportedConfigVariableNameRegex := regexp.MustCompile(`export\s+const\s+(\w+)\s*=\s*\w+\(`)

		// This regex matches the function name of an exported
		// config. For example, it'd match "cloudflareKv" in:
		// export const coreBaseKv = cloudflareKv({
		//   name: "CORE_BASE_KV",
		// } as const)
		functionNameRegex := regexp.MustCompile(`\s*=\s*(\w+)\(`)

		for _, possibleExportedConfig := range possibleExportedConfigs {
			possibleExportedConfigFunctionName := functionNameRegex.FindStringSubmatch(possibleExportedConfig)[1]

			// If possible exported config function name is equal to the
			// config setter function name, then the possible exported
			// config function name and possible exported config are
			// confirmed to represent actual configs.
			if possibleExportedConfigFunctionName == configSetterFunctionName {
				result[name] = &configData{
					variableName: possibleExportedConfigVariableNameRegex.FindStringSubmatch(possibleExportedConfig)[1],
					functionName: possibleExportedConfigFunctionName,
					exportString: possibleExportedConfig,
				}
				break
			}
		}
	}
	return result, nil
}

type nodeJsConfigScript = string

func (r *Resources) setNodeJsConfigScript(nameToConfigData nameToConfigData, groupToDepthToNames graph.GroupToDepthToNodes) (nodeJsConfigScript, error) {
	var functionNames []string

	functionNameToTrue := make(map[string]bool)
	for _, configData := range nameToConfigData {
		functionNameToTrue[configData.functionName] = true
		functionNames = append(functionNames, configData.functionName)
	}

	result := "import {\n"
	result += strings.Join(functionNames, ",\n")
	result += "\n} "
	result += "from \"@gasoline-dev/resources\"\n"

	// Configs have to be written in bottom-up dependency order to
	// avoid Node.js "cannot access 'variable name' before
	// initialization" errors. For example, given a graph of A->B,
	// B's config has to be written before A's because A will
	// reference B's config.
	for group := range groupToDepthToNames {
		numOfDepths := len(groupToDepthToNames[group])
		for depth := numOfDepths; depth >= 0; depth-- {
			for _, name := range groupToDepthToNames[group][depth] {
				result += strings.Replace(nameToConfigData[name].exportString, " as const", "", 1)
				result += "\n"
			}
		}
	}

	result += "const resourceNameToConfig = {}\n"

	for name, configData := range nameToConfigData {
		result += fmt.Sprintf("resourceNameToConfig[\"%s\"] = %s\n", name, configData.variableName)
	}

	result += "console.log(JSON.stringify(resourceNameToConfig))\n"

	return result, nil
}

type runNodeJsConfigScriptResult map[string]interface{}

func (r *Resources) runNodeJsConfigScript(nodeJsConfigScript nodeJsConfigScript) (runNodeJsConfigScriptResult, error) {
	cmd := exec.Command("node", "--input-type=module")

	cmd.Stdin = bytes.NewReader([]byte(nodeJsConfigScript))

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("unable to execute Node.js config script: %s\n%v", string(output), err)
	}

	strOutput := strings.TrimSpace(string(output))

	var result runNodeJsConfigScriptResult
	err = json.Unmarshal([]byte(strOutput), &result)
	if err != nil {
		return nil, fmt.Errorf("unable to unmarshal Node.js config script result: %v", err)
	}

	return result, nil
}

type nameToConfig = map[string]interface{}

func (r *Resources) setNameToConfig(runNodeJsConfigScriptResult runNodeJsConfigScriptResult) nameToConfig {
	result := make(nameToConfig)
	for name, config := range runNodeJsConfigScriptResult {
		c := config.(map[string]interface{})
		resourceType := c["type"].(string)
		result[name] = configs[resourceType](config.(map[string]interface{}))
	}
	return result
}

var configs = map[string]func(config config) interface{}{
	"cloudflare-kv": func(config config) interface{} {
		return &CloudflareKVConfig{
			ConfigCommon: ConfigCommon{
				Type: config["type"].(string),
				Name: config["name"].(string),
			},
		}
	},
}

type config map[string]interface{}

type ConfigCommon struct {
	Type string `json:"type"`
	Name string `json:"name"`
}

type CloudflareKVConfig struct {
	ConfigCommon
}

type CloudflareWorkerConfig struct {
	ConfigCommon
	KV []struct {
		Binding string `json:"binding"`
	} `json:"kv"`
}

File: packages/cli/src/modules/resources.ts
import { spawn } from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";
import type { ResourceNameToUpOutput } from "../cmds/up.js";
import {
	type GraphDepthToNodes,
	type GraphGroupToDepthToNodes,
	type GraphGroupToNodes,
	type GraphNodeToDepth,
	type GraphNodeToGroup,
	type GraphNodeToInDegrees,
	type GraphNodeToIntermediates,
	type GraphNodesWithInDegreesOfZero,
	setGraph,
} from "./graph.js";
import { deepMergeObjects } from "./objects.js";
import { convertKebabCaseToCapitalSnakeCase } from "./strings.js";
import type { TurboSummary } from "./turbo.js";

//
// Main resource setters.
//

type PackageJson = Record<string, unknown>;

export type ResourceContainerDirPath = string;

export type ResourceContainerSubdirPaths = string[];

async function setContainerSubdirPaths(
	containerDirPath: ResourceContainerDirPath,
): Promise<ResourceContainerSubdirPaths> {
	try {
		const entries = await fs.readdir(containerDirPath, { withFileTypes: true });
		return entries
			.filter((entry) => entry.isDirectory())
			.map((entry) => path.join(containerDirPath, entry.name));
	} catch (err) {
		throw new Error(
			`Unable to read resource container dir ${containerDirPath}: ${(err as Error).message}`,
		);
	}
}

export type ResourceNameToPackageJson = Record<string, PackageJson>;

function convertContainerSubdirPathToName(subdirPath: string): string {
	const subdirName = path.basename(subdirPath);
	return subdirName.replace(/-/g, "_").toUpperCase();
}

async function setNameToPackageJson(
	containerSubdirPaths: ResourceContainerSubdirPaths,
): Promise<ResourceNameToPackageJson> {
	const res: ResourceNameToPackageJson = {};
	for (const subdirPath of containerSubdirPaths) {
		const resourceName = convertContainerSubdirPathToName(subdirPath);
		const packageJsonPath = path.join(subdirPath, "package.json");

		try {
			const data = await fs.readFile(packageJsonPath, "utf8");
			const packageJson: PackageJson = JSON.parse(data);
			res[resourceName] = packageJson;
		} catch (err) {
			throw new Error(`Unable to read or parse ${packageJsonPath}: ${err}`);
		}
	}
	return res;
}

export type ResourceNameToDependencies = Record<string, string[]>;

function setNameToDependencies(
	nameToPackageJson: ResourceNameToPackageJson,
): ResourceNameToDependencies {
	const res: ResourceNameToDependencies = {};
	for (const name in nameToPackageJson) {
		const packageJson = nameToPackageJson[name];
		const resourceDependencies: string[] = [];
		const packageJsonDependencies = packageJson.dependencies as
			| Record<string, string>
			| undefined;
		if (packageJsonDependencies) {
			for (const packageJsonDependency in packageJsonDependencies) {
				const capitalSnakeCasePackageJsonDependency =
					convertKebabCaseToCapitalSnakeCase(packageJsonDependency);
				const internalDependency =
					nameToPackageJson[capitalSnakeCasePackageJsonDependency];
				if (internalDependency) {
					resourceDependencies.push(capitalSnakeCasePackageJsonDependency);
				}
			}
		}
		res[name] = resourceDependencies;
	}
	return res;
}

export type ResourceNameToFiles = {
	[name: string]: {
		configPath: string;
		buildPath: string;
		entityGroup: string;
		entity: string;
		cloud: string;
		cloudService: string;
		descriptor: string;
	};
};

async function setNameToFiles(
	containerDirPath: ResourceContainerDirPath,
	containerSubdirPaths: ResourceContainerSubdirPaths,
): Promise<ResourceNameToFiles> {
	const res: ResourceNameToFiles = {};

	const configPathPattern = /^index\.[^.]+\.[^.]+\.[^.]+\.[^.]+\.[^.]+\.ts$/;

	for (const subdirPath of containerSubdirPaths) {
		const resourceName = convertContainerSubdirPathToName(subdirPath);
		const srcPath = path.join(subdirPath, "src");

		try {
			const srcStats = await fs.stat(srcPath);
			if (!srcStats.isDirectory()) continue;

			const files = await fs.readdir(srcPath);
			for (const file of files) {
				if (configPathPattern.test(file)) {
					const configPath = path.join(srcPath, file);
					const relativePath = path.relative(containerDirPath, configPath);
					const parts = relativePath.split(path.sep);
					parts.splice(1, 0, "build");
					const buildPath = path
						.join(containerDirPath, ...parts)
						.replace(/\.ts$/, ".js");

					const [entityGroup, entity, cloud, cloudService, descriptor] = path
						.basename(subdirPath)
						.split("-");

					res[resourceName] = {
						configPath,
						buildPath,
						entityGroup,
						entity,
						cloud,
						cloudService,
						descriptor,
					};
					break;
				}
			}
		} catch (err) {
			if ((err as NodeJS.ErrnoException).code !== "ENOENT") {
				throw new Error(
					`Unable to read src dir ${srcPath}: ${(err as Error).message}`,
				);
			}
		}
	}
	return res;
}

export type ResourceNameToConfigFileContent = Record<string, string>;

async function setNameToConfigFileContent(
	nameToFiles: ResourceNameToFiles,
): Promise<ResourceNameToConfigFileContent> {
	const res: ResourceNameToConfigFileContent = {};
	for (const name in nameToFiles) {
		const configPath = nameToFiles[name].configPath;
		try {
			const content = await fs.readFile(configPath, "utf8");
			res[name] = content;
		} catch (err) {
			throw new Error(
				`Unable to read file ${configPath}: ${(err as Error).message}`,
			);
		}
	}
	return res;
}

type ConfigData = {
	variableName: string;
	functionName: string;
	exportString: string;
};

export type ResourceNameToConfigData = Record<string, ConfigData>;

function setNameToConfigData(
	nameToConfigFileContent: ResourceNameToConfigFileContent,
): ResourceNameToConfigData {
	const res: ResourceNameToConfigData = {};
	for (const name in nameToConfigFileContent) {
		const configFileContent = nameToConfigFileContent[name];
		const configSetterFunctionNameRegex =
			/import\s+\{[^}]*\b([a-z]+[A-Z][a-zA-Z]*)\b[^}]*\}\s+from\s+['"]@gasdotdev\/resources['"]/;
		const configSetterFunctionNameMatch = configFileContent.match(
			configSetterFunctionNameRegex,
		);
		if (!configSetterFunctionNameMatch) continue;
		const configSetterFunctionName = configSetterFunctionNameMatch[1];

		const exportedConfigRegex =
			/export\s+const\s+\w+\s*=\s*\w+\([\s\S]*?\)\s*(?:as\s*const\s*)?;?/gm;
		const possibleExportedConfigs =
			configFileContent.match(exportedConfigRegex) || [];

		const possibleExportedConfigVariableNameRegex =
			/export\s+const\s+(\w+)\s*=\s*\w+\(/;
		const functionNameRegex = /\s*=\s*(\w+)\(/;

		for (const possibleExportedConfig of possibleExportedConfigs) {
			const functionNameMatch = possibleExportedConfig.match(functionNameRegex);
			if (!functionNameMatch) continue;
			const possibleExportedConfigFunctionName = functionNameMatch[1];

			if (possibleExportedConfigFunctionName === configSetterFunctionName) {
				const variableNameMatch = possibleExportedConfig.match(
					possibleExportedConfigVariableNameRegex,
				);
				if (!variableNameMatch) continue;

				res[name] = {
					variableName: variableNameMatch[1],
					functionName: possibleExportedConfigFunctionName,
					exportString: possibleExportedConfig,
				};
				break;
			}
		}
	}
	return res;
}

export type ResourceNodeJsConfigScript = string;

function setNodeJsConfigScript(
	nameToConfigData: ResourceNameToConfigData,
	groupToDepthToNames: GraphGroupToDepthToNodes,
): ResourceNodeJsConfigScript {
	if (Object.keys(groupToDepthToNames).length === 0) {
		return "";
	}

	const functionNamesSet = new Set<string>();
	const functionNames: string[] = [];

	for (const name in nameToConfigData) {
		const functionName = nameToConfigData[name].functionName;
		if (!functionNamesSet.has(functionName)) {
			functionNamesSet.add(functionName);
			functionNames.push(functionName);
		}
	}

	let script = `import {\n${functionNames.join(",\n")}\n} from "@gasdotdev/resources"\n`;

	for (const group of Object.keys(groupToDepthToNames).map(Number)) {
		const numOfDepths = Object.keys(groupToDepthToNames[group]).length;
		for (let depth = numOfDepths - 1; depth >= 0; depth--) {
			for (const name of groupToDepthToNames[group][depth] || []) {
				if (nameToConfigData[name]) {
					script += `${nameToConfigData[name].exportString.replace(" as const", "")}\n`;
				}
			}
		}
	}

	script += "const resourceNameToConfig = {};\n";
	for (const name in nameToConfigData) {
		const configData = nameToConfigData[name];
		script += `resourceNameToConfig["${name}"] = ${configData.variableName};\n`;
	}
	script += "console.log(JSON.stringify(resourceNameToConfig));\n";

	return script;
}

type ResourceConfig = Record<string, any>;

export type ResourceNameToConfig = Record<string, ResourceConfig>;

function setNameToConfig(
	runNodeJsConfigScriptResult: ResourceRunNodeJsConfigScriptResult,
): ResourceNameToConfig {
	return { ...runNodeJsConfigScriptResult };
}

export type ResourceRunNodeJsConfigScriptResult = Record<
	string,
	ResourceConfig
>;

async function runNodeJsConfigScript(
	nodeJsConfigScript: ResourceNodeJsConfigScript,
): Promise<ResourceRunNodeJsConfigScriptResult> {
	if (nodeJsConfigScript === "") {
		return {};
	}

	return new Promise((resolve, reject) => {
		const child = spawn("node", ["--input-type=module"], {
			stdio: ["pipe", "pipe", "pipe"],
		});
		let stdout = "";
		let stderr = "";

		child.stdout.on("data", (data) => {
			stdout += data.toString();
		});
		child.stderr.on("data", (data) => {
			stderr += data.toString();
		});

		child.on("close", (code) => {
			if (code !== 0) {
				reject(new Error(`Node.js script execution error: ${stderr}`));
			} else {
				try {
					resolve(JSON.parse(stdout.trim()));
				} catch (err) {
					reject(
						new Error(
							`Unable to parse Node.js config script result: ${(err as Error).message}`,
						),
					);
				}
			}
		});

		child.stdin.write(nodeJsConfigScript);
		child.stdin.end();
	});
}

export type ResourceGroupToDepthToNames = GraphGroupToDepthToNodes;

export type Resources = {
	containerDirPath: ResourceContainerDirPath;
	containerSubdirPaths: ResourceContainerSubdirPaths;
	nameToPackageJson: ResourceNameToPackageJson;
	nameToFiles: ResourceNameToFiles;
	nameToConfigFileContent: ResourceNameToConfigFileContent;
	nameToConfigData: ResourceNameToConfigData;
	nameToDependencies: ResourceNameToDependencies;
	nameToIndegrees: GraphNodeToInDegrees;
	namesWithIndegreesOfZero: GraphNodesWithInDegreesOfZero;
	nameToIntermediates: GraphNodeToIntermediates;
	nameToGroup: GraphNodeToGroup;
	groupToNames: GraphGroupToNodes;
	depthToNames: GraphDepthToNodes;
	nameToDepth: GraphNodeToDepth;
	groupToDepthToNames: ResourceGroupToDepthToNames;
	nodeJsConfigScript: ResourceNodeJsConfigScript;
	runNodeJsConfigScriptResult: ResourceRunNodeJsConfigScriptResult;
	nameToConfig: ResourceNameToConfig;
};

type FactoryOptions = {
	upResources: UpResources;
};

/**
 *
 * @param options.upResources - Used in deployments to account for deleted resources.
 * Deleted resources are present in `gas.up.json` but not the resource container dir.
 * For example, if Resource X depends on Resource Y and Resource Y gets deleted,
 * Resource Y would no longer exist in the resource container dir. The only record of
 * Resource Y would be in `gas.up.json`. Therefore, the resource dependency data from
 * both are merged before graphing the resources. Only then, when all resources have
 * been accounted for, can a sufficient graph be generated for deployment.
 */
async function factory(
	containerDirPath: ResourceContainerDirPath,
	options?: FactoryOptions,
): Promise<Resources> {
	const containerSubdirPaths = await setContainerSubdirPaths(containerDirPath);

	const nameToPackageJson = await setNameToPackageJson(containerSubdirPaths);

	const nameToFiles = await setNameToFiles(
		containerDirPath,
		containerSubdirPaths,
	);

	const nameToConfigFileContent = await setNameToConfigFileContent(nameToFiles);

	const nameToConfigData = setNameToConfigData(nameToConfigFileContent);

	let nameToDependencies = setNameToDependencies(nameToPackageJson);

	if (options?.upResources) {
		const upResourceNameToDependencies: ResourceNameToDependencies = {};

		for (const name in options.upResources) {
			upResourceNameToDependencies[name] =
				options.upResources[name].dependencies;
		}

		nameToDependencies = {
			...upResourceNameToDependencies,
			...nameToDependencies,
		};
	}

	const graph = setGraph(nameToDependencies);

	const nameToIndegrees = graph.nodeToInDegrees;

	const namesWithIndegreesOfZero = graph.nodesWithInDegreesOfZero;

	const nameToIntermediates = graph.nodeToIntermediates;

	const nameToGroup = graph.nodeToGroup;

	const groupToNames = graph.groupToNodes;

	const depthToNames = graph.depthToNodes;

	const nameToDepth = graph.nodeToDepth;

	const groupToDepthToNames = graph.groupToDepthToNodes;

	const nodeJsConfigScript = setNodeJsConfigScript(
		nameToConfigData,
		groupToDepthToNames,
	);

	const runNodeJsConfigScriptResult =
		await runNodeJsConfigScript(nodeJsConfigScript);

	const nameToConfig = setNameToConfig(runNodeJsConfigScriptResult);

	return {
		containerDirPath,
		containerSubdirPaths,
		nameToPackageJson,
		nameToDependencies,
		nameToIndegrees,
		namesWithIndegreesOfZero,
		nameToIntermediates,
		nameToGroup,
		groupToNames,
		groupToDepthToNames,
		nameToFiles,
		nameToConfigFileContent,
		nameToConfigData,
		depthToNames,
		nameToDepth,
		nodeJsConfigScript,
		runNodeJsConfigScriptResult,
		nameToConfig,
	};
}

async function init(
	containerDirPath: ResourceContainerDirPath,
): Promise<Resources> {
	return await factory(containerDirPath);
}

export async function setResources(
	containerDirPath: ResourceContainerDirPath,
): Promise<Resources> {
	return await init(containerDirPath);
}

//
// Main resource helpers.
//
export type ResourceEntityGroups = string[];

export function setResourceEntityGroups(
	nameToFiles: ResourceNameToFiles,
	descriptorFilters?: string[],
): ResourceEntityGroups {
	const entityGroupSet = new Set<string>();
	for (const name in nameToFiles) {
		const files = nameToFiles[name];
		if (
			(!descriptorFilters ||
				descriptorFilters.length === 0 ||
				descriptorFilters.includes(files.descriptor)) &&
			!entityGroupSet.has(files.entityGroup)
		) {
			entityGroupSet.add(files.entityGroup);
		}
	}
	return Array.from(entityGroupSet);
}

export type ResourceEntities = string[];

export function setResourceEntities(
	nameToFiles: ResourceNameToFiles,
	descriptorFilters?: string[],
): ResourceEntities {
	const entitySet = new Set<string>();
	for (const name in nameToFiles) {
		const files = nameToFiles[name];
		if (
			(!descriptorFilters ||
				descriptorFilters.length === 0 ||
				descriptorFilters.includes(files.descriptor)) &&
			!entitySet.has(files.entity)
		) {
			entitySet.add(files.entity);
		}
	}
	return Array.from(entitySet);
}

//
// Up resource setters.
//

export type UpResourceConfig = Record<string, unknown>;

export type UpResourceDependencies = string[];

export type UpResourceOutput = Record<string, unknown>;

export type UpResources = {
	[name: string]: {
		config: UpResourceConfig;
		dependencies: UpResourceDependencies;
		output: UpResourceOutput;
	};
};

export async function setUpResources(path: string): Promise<UpResources> {
	const data = await fs.readFile(path, "utf8");
	return JSON.parse(data) as UpResources;
}

export type ResourceState = "CREATED" | "DELETED" | "UNCHANGED" | "UPDATED";

export type ResourceNameToState = {
	[name: string]: ResourceState;
};

function setNameToState(
	nameToConfig: ResourceNameToConfig,
	nameToDependencies: ResourceNameToDependencies,
	upResources: UpResources,
): ResourceNameToState {
	const nameToState: ResourceNameToState = {};

	for (const name in upResources) {
		if (!(name in nameToConfig)) {
			nameToState[name] = "DELETED";
		}
	}

	for (const name in nameToConfig) {
		if (!(name in upResources)) {
			nameToState[name] = "CREATED";
		} else {
			if (
				JSON.stringify(nameToConfig[name]) !==
				JSON.stringify(upResources[name].config)
			) {
				nameToState[name] = "UPDATED";
				continue;
			}

			if (
				JSON.stringify(nameToDependencies[name]) !==
				JSON.stringify(upResources[name].dependencies)
			) {
				nameToState[name] = "UPDATED";
				continue;
			}

			nameToState[name] = "UNCHANGED";
		}
	}

	return nameToState;
}

export type ResourceDeployState =
	| "CANCELED"
	| "CREATE_COMPLETE"
	| "CREATE_FAILED"
	| "CREATE_IN_PROGRESS"
	| "DELETE_COMPLETE"
	| "DELETE_FAILED"
	| "DELETE_IN_PROGRESS"
	| "PENDING"
	| "UPDATE_COMPLETE"
	| "UPDATE_FAILED"
	| "UPDATE_IN_PROGRESS";

export type ResourceNameToDeployState = {
	[name: string]: ResourceState | ResourceDeployState;
};

type GroupsWithStateChanges = number[];

function setGroupsWithStateChanges(
	nameToState: ResourceNameToState,
	nameToGroup: GraphNodeToGroup,
): GroupsWithStateChanges {
	const res: GroupsWithStateChanges = [];
	const seenGroups = new Set<number>();
	for (const name in nameToState) {
		if (nameToState[name] !== "UNCHANGED") {
			const group = nameToGroup[name];
			if (!seenGroups.has(group)) {
				res.push(group);
				seenGroups.add(group);
			}
		}
	}
	return res;
}

type GroupToHighestDeployDepth = {
	[group: number]: number;
};

function setGroupToHighestDeployDepth(
	groupsWithStateChanges: GroupsWithStateChanges,
	groupToNames: GraphGroupToNodes,
	nameToState: ResourceNameToState,
	nameToDepth: GraphNodeToDepth,
): GroupToHighestDeployDepth {
	const res: GroupToHighestDeployDepth = {};

	for (const group of groupsWithStateChanges) {
		let deployDepth = 0;
		let isFirstResourceToProcess = true;

		for (const name of groupToNames[group]) {
			// UNCHANGED resources aren't deployed, so its depth
			// can't be the deploy depth.
			if (nameToState[name] === "UNCHANGED") {
				continue;
			}

			// If resource is first to make it this far set deploy
			// depth so it can be used for comparison in future loops.
			if (isFirstResourceToProcess) {
				res[group] = nameToDepth[name];
				deployDepth = nameToDepth[name];
				isFirstResourceToProcess = false;
				continue;
			}

			// Update deploy depth if resource's depth is greater than
			// the comparative deploy depth.
			if (nameToDepth[name] > deployDepth) {
				res[group] = nameToDepth[name];
				deployDepth = nameToDepth[name];
			}
		}
	}

	return res;
}

export type ResourcesWithUp = Resources & {
	upResources: UpResources;
	nameToState: ResourceNameToState;
	nameToDeployState: ResourceNameToDeployState;
	groupToHighestDeployDepth: GroupToHighestDeployDepth;
	groupsWithStateChanges: GroupsWithStateChanges;
};

async function initWithUp(
	containerDirPath: ResourceContainerDirPath,
	upJsonPath: string,
): Promise<ResourcesWithUp> {
	const upResources = await setUpResources(upJsonPath);

	const resources = await factory(containerDirPath, {
		upResources,
	});

	const nameToState = setNameToState(
		resources.nameToConfig,
		resources.nameToDependencies,
		upResources,
	);

	const groupsWithStateChanges = setGroupsWithStateChanges(
		nameToState,
		resources.nameToGroup,
	);

	const groupToHighestDeployDepth = setGroupToHighestDeployDepth(
		groupsWithStateChanges,
		resources.groupToNames,
		nameToState,
		resources.nameToDepth,
	);

	return {
		...resources,
		upResources,
		nameToState,
		nameToDeployState: { ...nameToState },
		groupToHighestDeployDepth,
		groupsWithStateChanges,
	};
}

export async function setResourcesWithUp(
	containerDirPath: ResourceContainerDirPath,
	upJsonPath: string,
): Promise<ResourcesWithUp> {
	return await initWithUp(containerDirPath, upJsonPath);
}

//
// Up resource helpers.
//

export function setPostDeployUpResources(
	preDeployUpResources: UpResources,
	nameToConfig: ResourceNameToConfig,
	nameToDependencies: ResourceNameToDependencies,
	nameToUpOutput: ResourceNameToUpOutput,
	turboSummary: TurboSummary,
): UpResources {
	const newUpResources: UpResources = {};
	for (const name in nameToUpOutput) {
		newUpResources[name] = {
			config: nameToConfig[name],
			dependencies: nameToDependencies[name],
			output: nameToUpOutput[name],
		};
	}

	return deepMergeObjects<UpResources>(preDeployUpResources, newUpResources);
}
